src/Microsoft.Extensions.Configuration.AzureAppConfiguration/DomainVerifier.cs

```diff
@@ -0,0 +1,145 @@
+        private const int TlsPort = 443;
+        private const string SubjectAltNameOid = "2.5.29.17";
+        private const string CommonName = "CN=";
+        private const string MultiDomainWildcard = "*.";
+
+        public static async Task<string> GetValidDomain(Uri originEndpoint, string srvHostName)
+        {
+            if (string.IsNullOrEmpty(srvHostName))
+            {
+                return string.Empty;
+            }
+
+            IEnumerable<string> validDomains = await GetValidDomainsFromTlsCert(srvHostName).ConfigureAwait(false);
+            
+            foreach (var domain in validDomains)
+            {
+                if (originEndpoint.DnsSafeHost.EndsWith(domain, StringComparison.OrdinalIgnoreCase))
+                {
+                    return domain;
+                }
+            }
+
+            return string.Empty;
+        }
+
+        private static async Task<List<string>> GetValidDomainsFromTlsCert(string endpoint)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(endpoint));
+
+            var validDomains = new List<string>();
+
+            X509Certificate2 cert = null;
+
+            using (var client = new TcpClient(endpoint, TlsPort))
+            using (var sslStream = new SslStream(client.GetStream(), leaveInnerStreamOpen: false))
+            {
+
+                // Initiate the connection, so it will download the server certificate
+                await sslStream.AuthenticateAsClientAsync(endpoint).ConfigureAwait(false);
+
+                var serverCertificate = sslStream.RemoteCertificate;
+
+                if (serverCertificate == null)
+                {
+                    return validDomains;
+                }
+
+                cert = new X509Certificate2(serverCertificate);
+            }
+
+            if (cert == null)
+            {
+                return validDomains;
+            }
+
+            // Get the Subject Alternative Name (SAN) extension
+            var sanExtension = cert.Extensions[SubjectAltNameOid];
+
+            if (sanExtension != null)
+            {
+                IEnumerable<string> formattedSanExtensions = sanExtension
+                    .Format(true)
+                    .Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
+
+                foreach (string formattedExtension in formattedSanExtensions)
+                {
+                    // Valid pattern should be "DNS Name=*.domain.com"
+                    string[] parts = formattedExtension.Split('=');
+
+                    if (parts.Length != 2)
+                    {
+                        continue;
+                    }
+
+                    string value = parts[1];
+
+                    // Skip non-multi domain
+                    if (!value.StartsWith(MultiDomainWildcard))
+                    {
+                        continue;
+                    }
+
+                    // .domain.com
+                    string domain = value.Substring(1);

```

Comment:
Should it be value.Substring(MultiDomainWildcard.Length); ?

Reply:
We use `EndWith()` to validate, don't want to pass the validation for domains like ".xxxazconfig.io".

Reply:
I put this comment because hardcoding the '1' here seems out of place given that the check for startswith uses MultiDomainWildCard. I think it would be a bit easier to follow if "*." was hardcoded into the StartsWith check to be consistent with the `1` being hardcoded here.

Reply:
Not bad :), updated.